### PR TITLE
MueLu: add DOF-based interface aggregation for surface-coupled problems

### PR DESCRIPTION
**Pull request by mayrmt**
@trilinos/muelu 

## Motivation

This PR mainly adds the `InterfaceAggregationFactory`, which implements a specialized aggregation of Lagrange multiplier unknowns in case of surface-coupled problems. In contrast to #6931, this variant
- allows for multiple dual DOFs per node
- takes an `Xpetra::Map` describing the primal interface DOFs as input.

Tests for the `InterfaceAggregationFactory` in case meshtying problems (with an exact zero block on the main diagonal) are added.

To deal with meshtying problems, where the linear operator is an exact saddle-point system with a zero block on the main diagonal, the `ZeroSubBlockAFactory` is added.

Futher changes within this PR:
- adding tests and documentation of various classes/factories, among them
  - `MueLu::MapTransferFactory`
  - `MueLu::TestTestHelpers` (tests for the `MueLu:TestHelpers` class)
  - `Xpetra::MatrixUtils`
- cleanup source code at various places
- adding documentation at various places

## Related Issues

* Follows #6931, #7194 #8102

## Stakeholder Feedback

This change is required multilevel block preconditioning for surface-coupled problems in saddle-point formulation, i.e. when the Lagrange multiplier unknowns to enforce the coupling constraints are kept in the system.

## Testing

Existing tests pass. New tests for `MueLu::InterfaceAggregationFactory`, `MueLu::MapTransferFactory`, `MueLu::ZeroSubBlockAFactory`, `Xpetra::MatrixUtils` have been added.

## Additional Information

@kliegeois I've slightly adapted the input to your implementation, such that we can use the same factory for both designs. To this end, the implementation of the node-based aggregation factory is moved to a subroutine within the `InterfaceAggregationFactory`. Of course, the test is still up and running.

### CODE REVIEW

**Comment by mayrmt** [comment]
@trilinos/muelu There's still one build error related to the scalar type `complex`. Here it is:
```
/scratch/trilinos/jenkins/ascic144/workspace/trilinos-folder/Trilinos_pullrequest_gcc_7.2.0_serial/Trilinos/packages/muelu/test/unit_tests/MapTransferFactory.cpp: In instantiation of ‘void MueLuTests::MapTransferFactory_TransferPartialMap1D_UnitTest<Scalar, LocalOrdinal, GlobalOrdinal, Node>::runUnitTestImpl(Teuchos::FancyOStream&, bool&) const [with Scalar = std::complex<double>; LocalOrdinal = int; GlobalOrdinal = long long int; Node = Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial>; Teuchos::FancyOStream = Teuchos::basic_FancyOStream<char>]’:
/scratch/trilinos/jenkins/ascic144/workspace/trilinos-folder/Trilinos_pullrequest_gcc_7.2.0_serial/Trilinos/packages/muelu/test/unit_tests/../..Utils/MueLu_ETI_4arg.hpp:42:3:   required from here
/scratch/trilinos/jenkins/ascic144/workspace/trilinos-folder/Trilinos_pullrequest_gcc_7.2.0_serial/Trilinos/packages/muelu/test/unit_tests/MapTransferFactory.cpp:657:35: error: no match for ‘operator>’ (operand types are ‘const std::complex<double>’ and ‘double’)
         if (coarseVecEntries[lid] > Teuchos::ScalarTraits<MagnitudeType>::zero())
```
I tried several variants to fix it, but have not been successful, yet. Any advice is greatly appreciated.

**Comment by kliegeois** [comment]
> @trilinos/muelu There's still one build error related to the scalar type `complex`. Here it is:
> 
> ```
> /scratch/trilinos/jenkins/ascic144/workspace/trilinos-folder/Trilinos_pullrequest_gcc_7.2.0_serial/Trilinos/packages/muelu/test/unit_tests/MapTransferFactory.cpp: In instantiation of ‘void MueLuTests::MapTransferFactory_TransferPartialMap1D_UnitTest<Scalar, LocalOrdinal, GlobalOrdinal, Node>::runUnitTestImpl(Teuchos::FancyOStream&, bool&) const [with Scalar = std::complex<double>; LocalOrdinal = int; GlobalOrdinal = long long int; Node = Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial>; Teuchos::FancyOStream = Teuchos::basic_FancyOStream<char>]’:
> /scratch/trilinos/jenkins/ascic144/workspace/trilinos-folder/Trilinos_pullrequest_gcc_7.2.0_serial/Trilinos/packages/muelu/test/unit_tests/../..Utils/MueLu_ETI_4arg.hpp:42:3:   required from here
> /scratch/trilinos/jenkins/ascic144/workspace/trilinos-folder/Trilinos_pullrequest_gcc_7.2.0_serial/Trilinos/packages/muelu/test/unit_tests/MapTransferFactory.cpp:657:35: error: no match for ‘operator>’ (operand types are ‘const std::complex<double>’ and ‘double’)
>          if (coarseVecEntries[lid] > Teuchos::ScalarTraits<MagnitudeType>::zero())
>              ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ```
> 
> I tried several variants to fix it, but have not been successful, yet. Any advice is greatly appreciated.

Thanks @mayrmt for working on this!

Have you tried to replace:

```
if (coarseVecEntries[lid] > Teuchos::ScalarTraits<MagnitudeType>::zero())
```

by 

```
if (Teuchos::ScalarTraits<Scalar>::magnitude(coarseVecEntries[lid]) > Teuchos::ScalarTraits<MagnitudeType>::zero())
```

?

**Comment by mayrmt** [comment]
@kliegeois Thanks for the hint. I haven't tried this one, yet. Let's give it a spin!

**Comment by kliegeois** [comment]
> @kliegeois Thanks for the hint. I haven't tried this one, yet. Let's give it a spin!

It seems that it helped a bit :)

File: packages/muelu/test/unit_tests/ZeroSubBlockAFactory.cpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    TEST_ASSERT(!subBlockAFactory.is_null());
+  }
+
+  TEUCHOS_UNIT_TEST_TEMPLATE_4_DECL(ZeroSubBlockAFactory, ExtractZeroMainDiagonalBlockFrom2x2SaddlePointSystem, Scalar, LocalOrdinal, GlobalOrdinal, Node)
+  {
+#   include "MueLu_UseShortNames.hpp"
+
+    using test_factory = TestHelpers::TestFactory<SC, LO, GO, NO>;
+
+    MUELU_TEST_ONLY_FOR(Xpetra::UseTpetra)
```
Comment (mayrmt) [inline packages/muelu/test/unit_tests/ZeroSubBlockAFactory.cpp:78]: @trilinos/muelu This test should work for Epetra with `LocalOrdinal = int` and `GlobalOrdinal = int`, but I couldn't figure out how to limit the scope to this scenario. Other ordinal types seem not to implement the `CrsMatrix::expertStaticFillComplete()`. However, I couldn't figure out, how to limit the Epetra-version of this test to this particular scenario. I'd appreciate any help or hints. Thanks!

If somebody can assist with limiting the scope of this test as outlined above, I'd

**Comment by mayrmt** [comment]
@trilinos/muelu As far as I can see, just the `Trilinos_pullrequest_intel_17.0.1` tests fails, yet not with an actual error, but rather some obscure memory issue. I have followed the instructions on debugging as given on the CDash report, though it didn't turn up anything. 

I don't have access to an intel compiler, so I can't reproduce this specific build locally.

Any ideas how to proceed?

**Comment by cgcgcg** [comment]
@mayrmt It looks like something nasty is happening with some RCP of a `std::vector<int>`. Does this PR touch anything like that? I believe that this can happen when you take a weak RCP of something and then deallocate the underlying object. It's odd that it only happens on the intel build, but you try to use the debugging instructions given there on another build. I'd assume that the order in which the RCPs are created doesn't change from build to build.

**Comment by mayrmt** [comment]
Closing in favor of PRs #8593, #8588, #8589, which contain essentially the same changes, but in a cleaner way.

**Pull request closed without merging**